tests_and_analysis/test/euphonic_test/test_powder.py

```diff
@@ -0,0 +1,266 @@
+import numpy as np
+import numpy.testing as npt
+import pytest
+
+from euphonic import ureg, Crystal
+from euphonic.powder import (_get_default_bins, _get_qpts_sphere,
+                             _qpts_cart_to_frac,
+                             sample_sphere_dos, sample_sphere_structure_factor)
+
+sampling_functions = {
+    'golden': 'euphonic.sampling.golden_sphere',
+    'sphere-projected-grid': 'euphonic.sampling.sphere_from_square_grid',
+    'spherical-polar-grid': 'euphonic.sampling.spherical_polar_grid',
+    'spherical-polar-improved': 'euphonic.sampling.spherical_polar_improved',
+    'random-sphere': 'euphonic.sampling.random_sphere'}
+
+
+@pytest.fixture
+def random_qpts_array():
+    return np.random.random((4, 3))
+
+
+@pytest.fixture(params=[True, False])
+def jitter(request):
+    return request.param
+
+
+@pytest.mark.unit
+@pytest.mark.parametrize('npts, sampling, sampling_args, sampling_kwargs',
+                         [(10, 'golden', (10,), {'jitter': None}),
+                          # 31 pts rounded up to 2N^2 -> 4 cols, 8 rows
+                          (31, 'sphere-projected-grid',
+                           (8, 4), {'jitter': None}),
+                          (32, 'sphere-projected-grid',
+                           (8, 4), {'jitter': None}),
+                          (31, 'spherical-polar-grid',
+                           (8, 4), {'jitter': None}),
+                          (13, 'spherical-polar-improved',
+                           (13,), {'jitter': None}),
+                          (7, 'random-sphere', (7,), dict())]
+                         )
+def test_get_qpts_sphere(mocker, random_qpts_array, jitter,
+                         npts, sampling, sampling_args, sampling_kwargs):
+    """Check that functions from euphonic.sampling are called as expected"""
+    mocked_sampling_function = mocker.patch(sampling_functions[sampling],
+                                            return_value=random_qpts_array,
+                                            autospec=True)
+
+    if 'jitter' in sampling_kwargs:
+        sampling_kwargs['jitter'] = jitter
+
+    npt.assert_almost_equal(
+        _get_qpts_sphere(npts, sampling=sampling, jitter=jitter),
+        random_qpts_array)
+
+    mocked_sampling_function.assert_called_with(*sampling_args,
+                                                **sampling_kwargs)
```

Comment:
I'm wondering if we should test private functions and implementation details such as testing that the sampling functions have been called, but at the same time just testing the output of functions such as `sample_sphere_dos` wouldn't tell us much about why its failing...

This could provide useful information and has already been written so can be left in, but something to think about

Reply:
Yeah, the tests are a bit abstract unfortunately. The trouble is that these objects interact with nearly every class in Euphonic, so without extensive mocking the test will fail whenever something goes wrong anywhere else in Euphonic.

We should have some kind of integration/regression test on some actual numbers as well, to catch numerical changes in future. But developing such a test wouldn't really have told me whether the module is working as it is supposed to!

Reply:
Marking as "resolved" as there is no action required